### PR TITLE
added new format writer

### PR DESCRIPTION
**Pull request by JensWehner**
Tests are still missing, but I think this works now.

### CODE REVIEW

File: applications/Preprocessor/output_impl.h [hunk 1]
```diff
@@ -102,11 +105,20 @@ void outputMesh(Mesh<dimension>& mesh,
     }
     std::ofstream output(outputFileName.getValue());
     output << std::hexfloat;
-    output << "mesh 1" << std::endl;
+    size_t version = 2;
```
Comment (Laar) [inline applications/Preprocessor/output_impl.h:108]: Can be const

File: applications/Preprocessor/output_impl.h [hunk 2]
```diff
@@ -102,11 +105,20 @@ void outputMesh(Mesh<dimension>& mesh,
...
+    output << "mesh " << version << std::endl;
     output << mesh.getNumberOfNodes() << " " << mesh.getNumberOfElements()
            << " " << dimension;
     printOtherEntityCounts(output, mesh, Detail::tag<dimension - 1>{});
     output << std::endl;
+
+    output << "zones\n";
+    output << mesh.getZoneNames().size() << "\n";
+    for (size_t i = 0; i < mesh.getZoneNames().size(); i++) {
+        output << i << " " << mesh.getZoneNames()[i] << "\n";
```
Comment (Laar) [inline applications/Preprocessor/output_impl.h:118]: I'm wondering why you output i here? I don't see it in the new spec.

File: applications/Preprocessor/mesh_impl.h [hunk 3]
```diff
@@ -382,7 +382,8 @@ template <std::size_t dimension>
 void Mesh<dimension>::addElement(std::vector<std::size_t> nodeCoordinateIDs,
                                  const std::string& zoneName) {
     std::size_t elementID = elementsList.size();
-    Element<dimension> newElement{this, elementID, 0};
+    size_t zone_id=getZoneId(zoneName);
+    Element<dimension> newElement{this, elementID, zone_id};
```
Comment (Laar) [inline applications/Preprocessor/mesh_impl.h:386]: Oops.

**Review by Laar — COMMENTED** [review]
Looks like a good start.

Thought, I'm wondering why this is going on the `rewrite-centaur-reader` branch. That does not have to do anything with the new intermediate mesh format.

**Comment by JensWehner** [comment]
it has the new meshsource, I maybe was to lazy to find the parent branch

File: applications/Preprocessor/output_impl.h (hunk 2 above)
Comment (JensWehner) [reply applications/Preprocessor/output_impl.h:118]: ahh I thought the `#` in the spec was the number. Oops

**Comment by Laar** [comment]
> it has the new meshsource, I maybe was to lazy to find the parent branch

But this does not use anything of the new meshsource.

File: applications/Preprocessor/output_impl.h [hunk 4]
```diff
@@ -165,6 +177,8 @@ void outputMesh(Mesh<dimension>& mesh,
         }
         output << std::endl;
     }
+
+    output << "elements\n";
     for (auto element : mesh.getElements()) {
```
Comment (Laar) [inline applications/Preprocessor/output_impl.h:182]: I don't see the output of which element belongs to which zone here

Comment (JensWehner) [reply applications/Preprocessor/output_impl.h:182]: argh I forgot that part.... oops

**Review by Laar — COMMENTED** [review]
Looks good to me.

Just a request, could you rebase this so that it goes into the more logical intermediate-mesh-with-zones branch? As the rewrite of the Centaur reader does not have significant priority for me.

**Comment by Laar** [comment]
@JensWehner If you can fix the formatting, then I'll merge it.

**Comment by Laar** [comment]
As this outputs meshes that we can't read yet. For that we need #109, so I think it is best if we merge this directly after that.

File: applications/Preprocessor/lib/output_impl.h [hunk 5]
```diff
@@ -102,11 +103,18 @@ void outputMesh(Mesh<dimension>& mesh,
...
     std::ofstream output(outputFileName.getValue());
     output << std::hexfloat;
-    output << "mesh 1" << std::endl;
+    size_t version = 2;
+    output << "mesh " << version << std::endl;
     output << mesh.getNumberOfNodes() << " " << mesh.getNumberOfElements()
            << " " << dimension;
     printOtherEntityCounts(output, mesh, Detail::tag<dimension - 1>{});
     output << std::endl;
+    output << "zones" << std::endl;
```
Comment (Laar) [inline applications/Preprocessor/lib/output_impl.h:112]: I think the number of zones should be included here. Something like the following:

```suggestion
    output << "zones" << std::endl;
    output << mesh.getZoneNames().size() << std::endl;
```

**Review by Laar — CHANGES_REQUESTED** [review]
I think I found a difference with the specification.

File: applications/Preprocessor/lib/output_impl.h [hunk 6]
```diff
@@ -72,6 +72,7 @@ void printOtherEntities(std::ofstream& output, const Mesh<dimension>& mesh,
                         tag<d>) {
     for (auto entity : mesh.template getEntities<d>()) {
         std::set<std::size_t> localPartitions;
+        output << "codim" << dimension - d << '\n';
```
Comment (Laar) [inline applications/Preprocessor/lib/output_impl.h:75]: Codim is now repeated for every edge/face.

```suggestion
    output << "codim" << dimension - d << '\n';
    for (auto entity : mesh.template getEntities<d>()) {
        std::set<std::size_t> localPartitions;
```

**Review by Laar — CHANGES_REQUESTED** [review]
Another problem that I spotted when looking at the actual output. (I must have been sleeping in the initial review).

File: applications/Preprocessor/lib/output_impl.h [hunk 7]
```diff
@@ -102,11 +103,18 @@ void outputMesh(Mesh<dimension>& mesh,
...
            << " " << dimension;
     printOtherEntityCounts(output, mesh, Detail::tag<dimension - 1>{});
     output << std::endl;
+    output << "zones" << std::endl;
+    for (const auto& zonename : mesh.getZoneNames()) {
+        std::cout << zonename << '\n';
+    }
+
+    output << "nodes" << std::endl;
     std::size_t reservedSpace =
```
Comment (Laar) [inline applications/Preprocessor/lib/output_impl.h:118]: The ordering of the output is incorrect here. It should have been:
 - Partition information (this is old lines 110-119)
 - Zone information

**Review by Laar — CHANGES_REQUESTED** [review]
A problem with the output ordering

**Review by Laar — APPROVED** [review]
With some additional fixes it seems to work now.

**Pull request merged**